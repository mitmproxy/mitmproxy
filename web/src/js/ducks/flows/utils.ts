import { Comparer } from "@reduxjs/toolkit";

type Item = { id: string };

export function buildIndex<T extends Item>(
    data: T[],
): { [id: string]: number } {
    return Object.fromEntries(data.map((f, i) => [f.id, i]));
}

export function buildLookup<T extends Item>(
    data: T[],
): { [id: string]: boolean } {
    return Object.fromEntries(data.map((f) => [f.id, true]));
}

export function withKeyRemoved<T extends object>(map: T, key: string): T {
    if (key in map) {
        map = { ...map };
        delete map[key];
    }
    return map;
}

export function removeViewItemAt<T extends Item>(
    prevView: T[],
    prevViewIndex: { [id: string]: number },
    pos: number,
) {
    // update data
    const view = prevView.toSpliced(pos, 1);

    // update index
    const _viewIndex = { ...prevViewIndex };
    delete _viewIndex[prevView[pos].id];
    for (let i = view.length - 1; i >= pos; i--) {
        _viewIndex[view[i].id] = i;
    }

    return { view, _viewIndex };
}

export function updateViewItem<T extends Item>(
    prevView: T[],
    prevViewIndex: { [id: string]: number },
    item: T,
    sort: Comparer<T>,
) {
    const view = [...prevView];
    const _viewIndex = { ...prevViewIndex };
    let pos = _viewIndex[item.id];
    view[pos] = item;
    while (pos + 1 < view.length && sort(view[pos], view[pos + 1]) > 0) {
        view[pos] = view[pos + 1];
        view[pos + 1] = item;
        _viewIndex[item.id] = pos + 1;
        _viewIndex[view[pos].id] = pos;
        ++pos;
    }
    while (pos > 0 && sort(view[pos], view[pos - 1]) < 0) {
        view[pos] = view[pos - 1];
        view[pos - 1] = item;
        _viewIndex[item.id] = pos - 1;
        _viewIndex[view[pos].id] = pos;
        --pos;
    }
    return { view, _viewIndex };
}

export function insertViewItem<T extends Item>(
    prevView: T[],
    prevViewIndex: { [id: string]: number },
    item: T,
    sort: Comparer<T>,
) {
    const pos = findInsertPos(prevView, item, sort);

    const view = prevView.toSpliced(pos, 0, item);

    const _viewIndex = { ...prevViewIndex };
    for (let i = view.length - 1; i >= pos; i--) {
        _viewIndex[view[i].id] = i;
    }

    return { view, _viewIndex };
}

/// Find the insertion position in a sorted array.
export function findInsertPos<T>(
    list: T[],
    item: T,
    sort: Comparer<T>,
): number {
    let low = 0,
        high = list.length;

    // fast path: insert at end
    if (sort(list[high - 1], item) <= 0) {
        return high;
    }

    while (low < high) {
        const middle = (low + high) >>> 1;
        if (sort(item, list[middle]) >= 0) {
            low = middle + 1;
        } else {
            high = middle;
        }
    }

    return low;
}
