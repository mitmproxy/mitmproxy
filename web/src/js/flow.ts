/*
Type definitions for flows. Sync with mitmproxy/tools/web/app.py.
 */
interface _Flow {
    id: string
    intercepted: boolean
    is_replay: "request" | "response" | undefined
    type: string
    modified: boolean
    marked: string
    comment: string
    timestamp_created: number
    client_conn: Client
    server_conn?: Server
    error?: Error
}

export type Flow = HTTPFlow | TCPFlow | DNSFlow;

export interface HTTPFlow extends _Flow {
    type: "http"
    request: HTTPRequest
    response?: HTTPResponse
    websocket?: WebSocketData
}

export interface TCPFlow extends _Flow {
    type: "tcp"
    messages_meta: MessagesMeta,
}

export interface Error {
    msg: string
    timestamp: number
}

export type Address = [string, number];

export interface Connection {
    id: string
    peername?: Address
    sockname?: Address

    tls_established: boolean
    cert?: Certificate
    sni?: string | boolean
    cipher?: string
    alpn?: string
    tls_version?: string

    timestamp_start?: number
    timestamp_tls_setup?: number
    timestamp_end?: number

}

export interface Client extends Connection {
    peername: Address
    sockname: Address
    timestamp_start: number
}

export interface Server extends Connection {
    address?: Address
    timestamp_tcp_setup?: number
}

export interface Certificate {
    keyinfo: [string, number]
    sha256: string,
    notbefore: number,
    notafter: number,
    serial: string,
    subject: [string, string][],
    issuer: [string, string][],
    altnames: string[],
}

export type HTTPHeader = [name: string, value: string]
export type HTTPHeaders = HTTPHeader[];

export interface HTTPMessage {
    http_version: string
    headers: HTTPHeaders
    trailers?: HTTPHeaders
    contentLength?: number
    contentHash?: string
    timestamp_start: number
    timestamp_end?: number
}

export interface HTTPRequest extends HTTPMessage {
    method: string
    scheme: string
    host: string
    port: number
    path: string
    pretty_host: string
}

export interface HTTPResponse extends HTTPMessage {
    status_code: number
    reason: string
}

export interface MessagesMeta {
        contentLength: number,
        count: number,
        timestamp_last?: number
}

export interface WebSocketData {
    messages_meta: MessagesMeta,
    closed_by_client?: boolean
    close_code?: number
    close_reason?: string
    timestamp_end?: number
}

export interface DNSQuestion {
    name: string
    type: string
    class: string
}

export interface DNSResourceRecord {
    name: string
    type: string
    class: string
    ttl: number
    data: string
}

export interface DNSMessage {
    id: number
    query: boolean
    op_code: string
    authoritative_answer: boolean
    truncation: boolean
    recursion_desired: boolean
    recursion_available: boolean
    response_code: string
    status_code: number
    questions: DNSQuestion[]
    answers: DNSResourceRecord[]
    authorities: DNSResourceRecord[]
    additionals: DNSResourceRecord[]
    size: number
    timestamp: number
}

export interface DNSFlow extends _Flow {
    type: "dns"
    request: DNSMessage
    response?: DNSMessage
}
