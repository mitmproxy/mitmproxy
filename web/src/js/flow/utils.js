import _ from "lodash"

var defaultPorts = {
    "http": 80,
    "https": 443
};

export var MessageUtils = {
    getContentType: function (message) {
        var ct = this.get_first_header(message, /^Content-Type$/i);
        if(ct){
            return ct.split(";")[0].trim();
        }
    },
    get_first_header: function (message, regex) {
        //FIXME: Cache Invalidation.
        if (!message._headerLookups)
            Object.defineProperty(message, "_headerLookups", {
                value: {},
                configurable: false,
                enumerable: false,
                writable: false
            });
        if (!(regex in message._headerLookups)) {
            var header;
            for (var i = 0; i < message.headers.length; i++) {
                if (!!message.headers[i][0].match(regex)) {
                    header = message.headers[i];
                    break;
                }
            }
            message._headerLookups[regex] = header ? header[1] : undefined;
        }
        return message._headerLookups[regex];
    },
    match_header: function (message, regex) {
        var headers = message.headers;
        var i = headers.length;
        while (i--) {
            if (regex.test(headers[i].join(" "))) {
                return headers[i];
            }
        }
        return false;
    },
    getContentURL: function (flow, message, view) {
        if (message === flow.request) {
            message = "request";
        } else if (message === flow.response) {
            message = "response";
        }
        return `./flows/${flow.id}/${message}/` + (view ? `content/${view}.json` : 'content.data');
    }
};

export var RequestUtils = _.extend(MessageUtils, {
    pretty_url: function (request) {
        var port = "";
        if (defaultPorts[request.scheme] !== request.port) {
            port = ":" + request.port;
        }
        return request.scheme + "://" + request.pretty_host + port + request.path;
    }
});

export var ResponseUtils = _.extend(MessageUtils, {});


var parseUrl_regex = /^(?:(https?):\/\/)?([^\/:]+)?(?::(\d+))?(\/.*)?$/i;
export var parseUrl = function (url) {
    //there are many correct ways to parse a URL,
    //however, a mitmproxy user may also wish to generate a not-so-correct URL. ;-)
    var parts = parseUrl_regex.exec(url);
    if(!parts){
        return false;
    }

    var scheme = parts[1],
        host = parts[2],
        port = parseInt(parts[3]),
        path = parts[4];
    if (scheme) {
        port = port || defaultPorts[scheme];
    }
    var ret = {};
    if (scheme) {
        ret.scheme = scheme;
    }
    if (host) {
        ret.host = host;
    }
    if (port) {
        ret.port = port;
    }
    if (path) {
        ret.path = path;
    }
    return ret;
};


var isValidHttpVersion_regex = /^HTTP\/\d+(\.\d+)*$/i;
export var isValidHttpVersion = function (httpVersion) {
    return isValidHttpVersion_regex.test(httpVersion);
};
