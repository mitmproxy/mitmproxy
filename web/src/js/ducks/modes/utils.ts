import { getSpec as getRegularSpec } from "../../modes/regular";
import { getSpec as getLocalSpec } from "../../modes/local";
import { getSpec as getWireguardSpec } from "../../modes/wireguard";
import { getSpec as getReverseSpec } from "../../modes/reverse";
import { getSpec as getTransparentSpec } from "../../modes/transparent";
import { getSpec as getSocksSpec } from "../../modes/socks";
import { getSpec as getUpstreamSpec } from "../../modes/upstream";
import { getSpec as getDnsSpec } from "../../modes/dns";
import { fetchApi } from "../../utils";
import { BackendState } from "../backendState";
import {
    ActionReducerMapBuilder,
    AsyncThunk,
    Draft,
    PayloadAction,
} from "@reduxjs/toolkit";
import { AppAsyncThunkConfig, createAppAsyncThunk } from "../hooks";
import { ModeState, parseSpec, RawSpecParts } from "../../modes";

export const isActiveMode = (state: ModeState): boolean => {
    return state.active && !state.error;
};

/**
 * Async thunk to update modes based on current UI state.
 */
export async function updateModes(_, thunkAPI) {
    const modes = thunkAPI.getState().modes;
    const activeModes: string[] = [
        ...modes.regular.filter(isActiveMode).map(getRegularSpec),
        ...modes.local.filter(isActiveMode).map(getLocalSpec),
        ...modes.wireguard.filter(isActiveMode).map(getWireguardSpec),
        ...modes.reverse.filter(isActiveMode).map(getReverseSpec),
        ...modes.transparent.filter(isActiveMode).map(getTransparentSpec),
        ...modes.socks.filter(isActiveMode).map(getSocksSpec),
        ...modes.upstream.filter(isActiveMode).map(getUpstreamSpec),
        ...modes.dns.filter(isActiveMode).map(getDnsSpec),
        //add new modes here
    ];
    const response = await fetchApi.put("/options", {
        mode: activeModes,
    });
    if (response.status === 200) {
        return;
    } else {
        throw new Error(await response.text());
    }
}

export function createModeUpdateThunk<T>(type: string) {
    return createAppAsyncThunk<void, { server: ModeState; value: T }>(
        type,
        updateModes,
    );
}

export function addSetter<M extends ModeState, Attr extends keyof Draft<M>>(
    builder: ActionReducerMapBuilder<M[]>,
    attribute: Attr,
    setThunk: AsyncThunk<
        void,
        { server: ModeState; value: Draft<M>[Attr] },
        AppAsyncThunkConfig
    >,
) {
    builder.addCase(setThunk.pending, (state, action) => {
        const { server, value } = action.meta.arg;
        const idx = state.findIndex((m) => m.ui_id === server.ui_id);
        if (idx >= 0) {
            state[idx][attribute] = value;
            state[idx].error = undefined;
        }
    });
    builder.addCase(setThunk.rejected, (state, action) => {
        const { server } = action.meta.arg;
        const idx = state.findIndex((m) => m.ui_id === server.ui_id);
        if (idx >= 0) {
            state[idx].error = action.error.message;
        }
    });
}

export function updateState<M extends ModeState>(
    type: string,
    specPartsToState: (p: RawSpecParts) => M,
) {
    return function reducer(
        state: M[],
        action: PayloadAction<Partial<BackendState>>,
    ) {
        if (action.payload.servers) {
            const activeSpecs = Object.values(action.payload.servers)
                .filter((server) => server.type === type)
                .map((server) => parseSpec(server.full_spec));
            if (activeSpecs.length > 0) {
                return activeSpecs.map(specPartsToState);
            } else {
                for (const mode of state) {
                    mode.active = false;
                }
            }
        }
    };
}
