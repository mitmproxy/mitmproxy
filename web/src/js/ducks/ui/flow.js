import * as flowsActions from '../flows'
import { getDiff } from "../../utils"

import _ from 'lodash'

export const SET_CONTENT_VIEW               = 'UI_FLOWVIEW_SET_CONTENT_VIEW',
             DISPLAY_LARGE                  = 'UI_FLOWVIEW_DISPLAY_LARGE',
             SET_TAB                        = "UI_FLOWVIEW_SET_TAB",
             START_EDIT                     = 'UI_FLOWVIEW_START_EDIT',
             UPDATE_EDIT                    = 'UI_FLOWVIEW_UPDATE_EDIT',
             UPLOAD_CONTENT                 = 'UI_FLOWVIEW_UPLOAD_CONTENT',
             SET_SHOW_FULL_CONTENT          = 'UI_SET_SHOW_FULL_CONTENT',
             SET_CONTENT_VIEW_DESCRIPTION   = "UI_SET_CONTENT_VIEW_DESCRIPTION",
             SET_CONTENT                    = "UI_SET_CONTENT"


const defaultState = {
    displayLarge: false,
    viewDescription: '',
    showFullContent: false,
    modifiedFlow: false,
    contentView: 'Auto',
    tab: 'request',
    content: [],
    maxContentLines: 80,
}

export default function reducer(state = defaultState, action) {
    let wasInEditMode = !!(state.modifiedFlow)

    let content = action.content || state.content
    let isFullContentShown = content && content.length <= state.maxContentLines

    switch (action.type) {

        case START_EDIT:
            return {
                ...state,
                modifiedFlow: action.flow,
                contentView: 'Edit',
                showFullContent: true
            }

        case UPDATE_EDIT:
            return {
                ...state,
                modifiedFlow: _.merge({}, state.modifiedFlow, action.update)
            }

        case flowsActions.SELECT:
            return {
                ...state,
                modifiedFlow: false,
                displayLarge: false,
                contentView: (wasInEditMode ? 'Auto' : state.contentView),
                showFullContent: isFullContentShown,
            }

        case flowsActions.UPDATE:
            // There is no explicit "stop edit" event.
            // We stop editing when we receive an update for
            // the currently edited flow from the server
            if (action.item.id === state.modifiedFlow.id) {
                return {
                    ...state,
                    modifiedFlow: false,
                    displayLarge: false,
                    contentView: (wasInEditMode ? 'Auto' : state.contentView),
                    showFullContent: false
                }
            } else {
                return state
            }

        case SET_CONTENT_VIEW_DESCRIPTION:
            return {
                ...state,
                viewDescription: action.description
            }

        case SET_SHOW_FULL_CONTENT:
            return {
                ...state,
                showFullContent: true
            }

        case SET_TAB:
            return {
                ...state,
                tab: action.tab ? action.tab : 'request',
                displayLarge: false,
                showFullContent: false
            }

        case SET_CONTENT_VIEW:
            return {
                ...state,
                contentView: action.contentView,
                showFullContent: action.contentView == 'Edit'
            }

        case SET_CONTENT:
            return {
                ...state,
                content: action.content,
                showFullContent: isFullContentShown
            }

        case DISPLAY_LARGE:
            return {
                ...state,
                displayLarge: true,
            }
        default:
            return state
    }
}

export function setContentView(contentView) {
    return { type: SET_CONTENT_VIEW, contentView }
}

export function displayLarge() {
    return { type: DISPLAY_LARGE }
}

export function selectTab(tab) {
    return { type: SET_TAB, tab }
}

export function startEdit(flow) {
    return { type: START_EDIT, flow }
}

export function updateEdit(update) {
    return { type: UPDATE_EDIT, update }
}

export function setContentViewDescription(description) {
    return { type: SET_CONTENT_VIEW_DESCRIPTION, description }
}

export function setShowFullContent() {
    return { type: SET_SHOW_FULL_CONTENT }
}

export function setContent(content){
    return { type: SET_CONTENT, content}
}

export function stopEdit(flow, modifiedFlow) {
    let diff = getDiff(flow, modifiedFlow)
    return flowsActions.update(flow, diff)
}
