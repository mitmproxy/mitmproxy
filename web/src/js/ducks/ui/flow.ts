import * as flowActions from "../flows";
import {tabsForFlow} from "../../components/FlowView";

export const
    SET_TAB = "UI_FLOWVIEW_SET_TAB",
    SET_CONTENT_VIEW_FOR = "SET_CONTENT_VIEW_FOR"


interface UiFlowState {
    tab: string
    contentViewFor: { [messageId: string]: string }
}

export const defaultState: UiFlowState = {
    tab: 'request',
    contentViewFor: {},
}

export default function reducer(state = defaultState, action): UiFlowState {
    switch (action.type) {

        case SET_CONTENT_VIEW_FOR:
            return {
                ...state,
                contentViewFor: {
                    ...state.contentViewFor,
                    [action.messageId]: action.contentView
                }
            }

        case SET_TAB:
            return {
                ...state,
                tab: action.tab ? action.tab : 'request',
            }

        default:
            return state
    }
}

export function selectTab(tab) {
    return {type: SET_TAB, tab}
}

export function setContentViewFor(messageId: string, contentView: string) {
    return {type: SET_CONTENT_VIEW_FOR, messageId, contentView}
}
