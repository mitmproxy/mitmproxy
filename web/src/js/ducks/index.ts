import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
