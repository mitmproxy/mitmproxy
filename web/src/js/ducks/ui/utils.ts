import { Flow } from "../../flow";

/**
 * mitmwebnext implementation.
 */
export function tabsForFlowNext(flow: Flow): {
    request: string[];
    response: string[];
} {
    let responseTabs: string[] = [],
        requestTabs: string[] = [];

    switch (flow.type) {
        case "http":
            responseTabs = [
                "headers",
                ...filterExistingProperties(flow, ["response", "websocket"]),
            ];
            requestTabs = ["headers", "query", "cookies", "body"];
            break;
        case "tcp":
            responseTabs = ["tcpmessages"];
            break;
        case "udp":
            responseTabs = ["udpmessages"];
            break;
        case "dns":
            responseTabs = filterExistingProperties(flow, [
                "request",
                "response",
            ]).map((s) => "dns" + s);
            break;
    }

    if (flow.error) {
        responseTabs.push("error");
    }

    responseTabs.push("connection");
    responseTabs.push("timing");
    responseTabs.push("comment");

    return { response: responseTabs, request: requestTabs };
}

/**
 * mitmweb implementation.
 */
export function tabsForFlow(flow: Flow): string[] {
    let tabs;
    switch (flow.type) {
        case "http":
            tabs = ["request", "response", "websocket"].filter((k) => flow[k]);
            break;
        case "tcp":
            tabs = ["tcpmessages"];
            break;
        case "udp":
            tabs = ["udpmessages"];
            break;
        case "dns":
            tabs = ["request", "response"]
                .filter((k) => flow[k])
                .map((s) => "dns" + s);
            break;
    }

    if (flow.error) tabs.push("error");
    tabs.push("connection");
    tabs.push("timing");
    tabs.push("comment");
    return tabs;
}

/**
 * The flow may not yet contain all properties if it hasn't been finalized yet.
 * This function filters them out.
 */
function filterExistingProperties(flow: Flow, tabs: string[]) {
    return tabs.filter((tab) => {
        return (flow as any)[tab] !== undefined;
    });
}
