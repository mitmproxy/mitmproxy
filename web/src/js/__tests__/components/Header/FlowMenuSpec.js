jest.mock('../../../flow/utils')

import React from 'react'
import renderer from 'react-test-renderer'
import ConnectedFlowMenu, { FlowMenu } from '../../../components/Header/FlowMenu'
import { TFlow, TStore }from '../../ducks/tutils'
import { MessageUtils } from "../../../flow/utils"
import { Provider } from 'react-redux'

describe('FlowMenu Component', () => {
    let actions = {
        resumeFlow: jest.fn(),
        killFlow: jest.fn(),
        replayFlow: jest.fn(),
        duplicateFlow: jest.fn(),
        removeFlow: jest.fn(),
        revertFlow: jest.fn()
    },
        tflow = new TFlow()
    tflow.modified = true
    tflow.intercepted = true

    it('should render correctly without flow', () => {
        let flowMenu = renderer.create(
                <FlowMenu removeFlow={actions.removeFlow}
                           killFlow={actions.killFlow}
                           replayFlow={actions.replayFlow}
                           duplicateFlow={actions.duplicateFlow}
                           resumeFlow={actions.resumeFlow}
                           revertFlow={actions.revertFlow}/>),
            tree = flowMenu.toJSON()
            expect(tree).toMatchSnapshot()
    })

    let flowMenu = renderer.create(<FlowMenu
                            flow={tflow}
                            removeFlow={actions.removeFlow}
                            killFlow={actions.killFlow}
                            replayFlow={actions.replayFlow}
                            duplicateFlow={actions.duplicateFlow}
                            resumeFlow={actions.resumeFlow}
                            revertFlow={actions.revertFlow}/>),
        tree = flowMenu.toJSON()

    it('should render correctly with flow', () => {
        expect(tree).toMatchSnapshot()
    })

    let menu_content_1 = tree.children[0].children[0]
    it('should handle replayFlow', () => {
        let button = menu_content_1.children[0]
        button.props.onClick()
        expect(actions.replayFlow).toBeCalledWith(tflow)
    })

    it('should handle duplicateFlow', () => {
        let button = menu_content_1.children[1]
        button.props.onClick()
        expect(actions.duplicateFlow).toBeCalledWith(tflow)
    })

    it('should handle revertFlow', () => {
        let button = menu_content_1.children[2]
        button.props.onClick()
        expect(actions.revertFlow).toBeCalledWith(tflow)
    })

    it('should handle removeFlow', () => {
        let button = menu_content_1.children[3]
        button.props.onClick()
        expect(actions.removeFlow).toBeCalledWith(tflow)
    })

    let menu_content_2 = tree.children[1].children[0]
    it('should handle download', () => {
        let button = menu_content_2.children[0]
        button.props.onClick()
        expect(MessageUtils.getContentURL).toBeCalledWith(tflow, tflow.response)
    })

    let menu_content_3 = tree.children[2].children[0]
    it('should handle resumeFlow', () => {
        let button = menu_content_3.children[0]
        button.props.onClick()
        expect(actions.resumeFlow).toBeCalledWith(tflow)
    })

    it('should handle killFlow', () => {
        let button = menu_content_3.children[1]
        button.props.onClick()
        expect(actions.killFlow).toBeCalledWith(tflow)
    })

    it('should connect to state', () => {
        let store = TStore(),
            provider = renderer.create(<Provider store={store}><ConnectedFlowMenu/></Provider>),
            tree = provider.toJSON()
        expect(tree).toMatchSnapshot()
    })

})
