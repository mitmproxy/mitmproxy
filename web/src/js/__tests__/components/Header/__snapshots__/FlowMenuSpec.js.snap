// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowMenu Component should connect to state 1`] = `
<div
  className="flow-menu"
>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[r]eplay flow"
      >
        <i
          className="fa fa-fw fa-repeat text-primary"
        />
        Replay
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[D]uplicate flow"
      >
        <i
          className="fa fa-fw fa-copy text-info"
        />
        Duplicate
      </button>
      <button
        className="btn btn-default"
        disabled={true}
        title="revert changes to flow [V]"
      >
        <i
          className="fa fa-fw fa-history text-warning"
        />
        Revert
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[d]elete flow"
      >
        <i
          className="fa fa-fw fa-trash text-danger"
        />
        Delete
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Flow Modification
    </div>
  </div>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="download"
      >
        <i
          className="fa fa-fw fa-download"
        />
        Download
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Export
    </div>
  </div>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        disabled={true}
        title="[a]ccept intercepted flow"
      >
        <i
          className="fa fa-fw fa-play text-success"
        />
        Resume
      </button>
      <button
        className="btn btn-default"
        disabled={true}
        title="kill intercepted flow [x]"
      >
        <i
          className="fa fa-fw fa-times text-danger"
        />
        Abort
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Interception
    </div>
  </div>
</div>
`;

exports[`FlowMenu Component should render correctly with flow 1`] = `
<div
  className="flow-menu"
>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[r]eplay flow"
      >
        <i
          className="fa fa-fw fa-repeat text-primary"
        />
        Replay
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[D]uplicate flow"
      >
        <i
          className="fa fa-fw fa-copy text-info"
        />
        Duplicate
      </button>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        title="revert changes to flow [V]"
      >
        <i
          className="fa fa-fw fa-history text-warning"
        />
        Revert
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="[d]elete flow"
      >
        <i
          className="fa fa-fw fa-trash text-danger"
        />
        Delete
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Flow Modification
    </div>
  </div>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        onClick={[Function]}
        title="download"
      >
        <i
          className="fa fa-fw fa-download"
        />
        Download
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Export
    </div>
  </div>
  <div
    className="menu-group"
  >
    <div
      className="menu-content"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        title="[a]ccept intercepted flow"
      >
        <i
          className="fa fa-fw fa-play text-success"
        />
        Resume
      </button>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        title="kill intercepted flow [x]"
      >
        <i
          className="fa fa-fw fa-times text-danger"
        />
        Abort
      </button>
    </div>
    <div
      className="menu-legend"
    >
      Interception
    </div>
  </div>
</div>
`;

exports[`FlowMenu Component should render correctly without flow 1`] = `<div />`;
